content/6_Advanced/SPneg.mdx

```diff
@@ -43,7 +43,99 @@ frequency: 0
 
 #### Solution
 
-<IncompleteSection />
+Although this is a Single Source Shortest Paths problem, we can not use the known
+Dijkstra's algorithm, because we have negative weights on edges. An alternative is
+to use the Bellman-Ford algorithm. The algorithm first considers all the paths
+which use 1 edge. Then it calculates all the paths with at most 2 edges, and so
+on. If the graph has no negative cycle, then the shortest path between the source
+and any other vertice should have at most $V-1$ edges, where $V$ stands for the number
+of vertices in the graph. Because of that, the algorithm iterates through at most
+all edges $V-1$ times. Hence, it runs in $O(E \cdot V)$.
+
+If the graph has a negative cycle, we can detect a vertice in this cycle by running
+another relaxation. In this problem, belonging to a negative cycle means the
+distance to that point is negative infinity. Notice that all points that are reachable
+from those will also have minus infinity cost. In the solution below, we detect
+all the negative cycles and save the point from which we detected the cycle. We then
+do a [BFS](/silver/graph-traversal) with those points as sources.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int INF = 1e9;
```

Comment:
why not INT32_MAX

Reply:
Avoid overflow when performing addition. For this implementation in particular it does not make a difference, but it is usually helpful that INF + INF does not overflow.

Reply:
so are you doing inf + inf’s here or not

Reply:
Not doing

Reply:
then just use int32_max

Reply:
Ok